### PR TITLE
2 file copyed from PR81

### PR DESCRIPTION
**Pull request by pradhyumnap-hub**
Fixes #<issue_number_goes_here>

> It's a good idea to open an issue first for discussion.

- [ ] Tests pass
- [ ] Appropriate changes to documentation are included in the PR

### CODE REVIEW

File: frontend/src/app/audio/audio.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+    fixture.detectChanges();
+
+    const lightboxElement = fixture.debugElement.query(
+      By.css('app-media-lightbox'),
+    );
+    expect(lightboxElement).toBeTruthy();
+    expect(component.mediaItem).toEqual(mockMediaItem);
+    flush();
+  }));
+});
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: changed in code

File: frontend/src/app/audio/audio.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+import { MatSelectChange, MatSelectModule } from '@angular/material/select';
+import { MatIconModule } from '@angular/material/icon';
+import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
+// Removed MediaLightboxComponent import - using mock instead
+import { HarnessLoader } from '@angular/cdk/testing';
+import { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';
+import { MatDividerModule } from '@angular/material/divider';
+import { LanguageEnum, VoiceEnum } from './audio.constants';
+import { By } from '@angular/platform-browser';
+import { AddVoiceDialogComponent } from '../components/add-voice-dialog/add-voice-dialog.component';
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: removed unused imports

File: frontend/src/app/audio/audio.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+    component = fixture.componentInstance;
+    dialog = TestBed.inject(MatDialog) as jasmine.SpyObj<MatDialog>;
+    notificationService = TestBed.inject(
+      NotificationService,
+    ) as jasmine.SpyObj<NotificationService>;
+    workspaceStateService = TestBed.inject(
+      WorkspaceStateService,
+    ) as jasmine.SpyObj<WorkspaceStateService>;
+    audioService = TestBed.inject(AudioService) as jasmine.SpyObj<AudioService>;
+    //loader = TestbedHarnessEnvironment.loader(fixture);
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: deleted in code

File: frontend/src/app/audio/audio.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+    }
+    return undefined;
+  }
+}
+describe('AudioComponent', () => {
+  let component: AudioComponent;
+  let fixture: ComponentFixture<AudioComponent>;
+  let workspaceStateService: jasmine.SpyObj<WorkspaceStateService>;
+  let dialog: jasmine.SpyObj<MatDialog>;
+  //let loader: HarnessLoader;
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: deleted in code

File: frontend/src/app/audio/audio.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+import { WorkspaceStateService } from '../services/workspace/workspace-state.service';
+import { MatSelectChange } from '@angular/material/select';
+import { MatIconModule } from '@angular/material/icon';
+import { MatFormFieldModule } from '@angular/material/form-field';
+import { MatInputModule } from '@angular/material/input';
+import { MatButtonModule } from '@angular/material/button';
+import { FormsModule } from '@angular/forms';
+// Removed MediaLightboxComponent import - using mock instead
+//import { HarnessLoader } from '@angular/cdk/testing';
+// { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: deleted in code

File: frontend/src/app/audio/audio.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+      expect(audioService.generateAudio).toHaveBeenCalled();
+      expect(component.isLoading).toBeFalse();
+      expect(notificationService.show).toHaveBeenCalledWith(
+        'Generation failed',
+        'error',
+        'cross-in-circle-white',
+        undefined,
+        20000,
+      );
+      tick(20000);
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: deleted in code

File: frontend/src/app/audio/audio.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+      component.generate();
+
+      expect(component.isLoading).toBeTrue();
+      expect(component.mediaItem).toBeNull();
+      expect(component.audioUrl).toBeNull();
+
+      audioSubject.next(mockMediaItem);
+      audioSubject.complete();
+      tick();
+
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: deleted in code

File: frontend/src/app/audio/audio.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+    }
+    return undefined;
+  }
+}
+describe('AudioComponent', () => {
+  let component: AudioComponent;
+  let fixture: ComponentFixture<AudioComponent>;
+  let workspaceStateService: jasmine.SpyObj<WorkspaceStateService>;
+  let dialog: jasmine.SpyObj<MatDialog>;
+  let loader: HarnessLoader;
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: removed in code

File: frontend/src/app/audio/audio.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+    component = fixture.componentInstance;
+    dialog = TestBed.inject(MatDialog) as jasmine.SpyObj<MatDialog>;
+    notificationService = TestBed.inject(
+      NotificationService,
+    ) as jasmine.SpyObj<NotificationService>;
+    workspaceStateService = TestBed.inject(
+      WorkspaceStateService,
+    ) as jasmine.SpyObj<WorkspaceStateService>;
+    audioService = TestBed.inject(AudioService) as jasmine.SpyObj<AudioService>;
+    loader = TestbedHarnessEnvironment.loader(fixture);
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: removed in code

File: frontend/src/app/audio/audio.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+    return this.mediaItem?.mimeType?.startsWith('audio/') ?? false;
+  }
+  get posterUrl(): string | undefined {
+    if (this.isVideo && this.mediaItem?.presignedThumbnailUrls?.length) {
+      return this.mediaItem.presignedThumbnailUrls[this.selectedIndex];
+    }
+    return undefined;
+  }
+}
+describe('AudioComponent', () => {
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: space added

File: frontend/src/app/audio/audio.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+import { WorkspaceStateService } from '../services/workspace/workspace-state.service';
+import { MatSelectChange } from '@angular/material/select';
+import { MatIconModule } from '@angular/material/icon';
+import { MatFormFieldModule } from '@angular/material/form-field';
+import { MatInputModule } from '@angular/material/input';
+import { MatButtonModule } from '@angular/material/button';
+import { FormsModule } from '@angular/forms';
+// Removed MediaLightboxComponent import - using mock instead
+import { HarnessLoader } from '@angular/cdk/testing';
+import { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: removed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,450 @@
...
+} from '../services/audio/audio.service';
+import { of, throwError, Subject } from 'rxjs';
+import { JobStatus, MediaItem } from '../common/models/media-item.model';
+import { WorkspaceStateService } from '../services/workspace/workspace-state.service';
+import { MatSelectChange } from '@angular/material/select';
+import { MatIconModule } from '@angular/material/icon';
+// import { MatFormFieldModule } from '@angular/material/form-field';
+// import { MatInputModule } from '@angular/material/input';
+// import { MatButtonModule } from '@angular/material/button';
+// import { FormsModule } from '@angular/forms';
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: commented code removed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 13]
```diff
@@ -0,0 +1,450 @@
...
+      declarations: [AudioComponent],
+      imports: [
+        HttpClientTestingModule,
+        MatIconModule,
+        MockMediaLightboxComponent,
+        // CommonModule,
+        // FormsModule,
+        // MatFormFieldModule,
+        // MatInputModule,
+        // MatButtonModule,
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: commented code removed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+      ['show']
+    );
+
+    const dialogSpy = jasmine.createSpyObj('MatDialog', ['open']);
+
+    await TestBed.configureTestingModule({
+      declarations: [AudioComponent],
+    await TestBed.configureTestingModule({
+      declarations: [AudioComponent],
+      imports: [
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: resolved in code

File: frontend/src/app/audio/audio.component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,440 @@
...
+  }>();
+  @Output() sendToVtoClicked = new EventEmitter<number>();
+  @Output() extendWithAiClicked = new EventEmitter<{
+    mediaItem: any;
+    selectedIndex: number;
+  }>();
+  @Output() concatenateClicked = new EventEmitter<{
+    mediaItem: any;
+    selectedIndex: number;
+  }>();
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts:85]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,440 @@
...
+      <img *ngIf="selectedUrl && !isAudio && !isVideo" [src]="selectedUrl" [alt]="mediaItem.originalPrompt" [width]="imageWidth" [height]="imageHeight" class="main-media" />
+      <video *ngIf="selectedUrl && isVideo" [src]="selectedUrl" [poster]="posterUrl" class="main-media" controls muted></video>
+      <audio *ngIf="selectedUrl && isAudio" [src]="selectedUrl" controls></audio>
+    </div>
+  `,
+  standalone: true, // Keep it standalone like the real one
+  imports: [CommonModule] // CommonModule for *ngIf
+})
+class MockMediaLightboxComponent {
+  @Input() mediaItem: any; // Use 'any' for simplicity in mock
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 17]
```diff
@@ -0,0 +1,440 @@
...
+  @Output() editClicked = new EventEmitter<number>();
+  @Output() generateVideoClicked = new EventEmitter<{
+    role: 'start' | 'end';
+    index: number;
+  }>();
+  @Output() sendToVtoClicked = new EventEmitter<number>();
+  @Output() extendWithAiClicked = new EventEmitter<{
+    mediaItem: any;
+    selectedIndex: number;
+  }>();
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts:81]: fixed

File: frontend/src/app/audio/audio.component.ts [hunk 18]
```diff
@@ -199,18 +197,26 @@ export class AudioComponent {
...
 
     const activeWorkspaceId = this.workspaceStateService.getActiveWorkspaceId();
     if (!activeWorkspaceId) {
-      handleErrorSnackbar(this.snackBar, { message: 'Please select a workspace first.' }, 'Workspace');
+      this.notificationService.show(
+        'Please select a workspace first.',
+        'error',
+        'cross-in-circle-white',
+        undefined,
+        20000
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.ts]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 19]
```diff
@@ -0,0 +1,440 @@
...
+import { MatSelectChange } from '@angular/material/select';
+import { MatIconModule } from '@angular/material/icon';
+import { LanguageEnum, VoiceEnum } from './audio.constants';
+import { By } from '@angular/platform-browser';
+import { ActivatedRoute } from '@angular/router';
+import { NotificationService } from '../common/services/notification.service';
+import { CommonModule } from '@angular/common'; // Needed for *ngIf in mock template
+
+
+// Define a mock MediaLightboxComponent
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts:50]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 20]
```diff
@@ -0,0 +1,440 @@
...
+        sampleCount: 2,
+        languageCode: undefined,
+        voiceName: undefined,
+      };
+
+      component.generate();
+
+      expect(audioService.generateAudio).toHaveBeenCalledWith(expectedRequest);
+    });
+    it('should call audioService with correct params for Chirp', () => {
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts:275]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 21]
```diff
@@ -0,0 +1,440 @@
...
+        'cross-in-circle-white',
+        undefined,
+        20000,
+      );
+    }));
+  });
+
+  describe('Audio Player', () => {
+    let audioEl: HTMLAudioElement;
+
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 22]
```diff
@@ -0,0 +1,439 @@
...
+      tick();
+      fixture.detectChanges();
+      expect(audioService.generateAudio).toHaveBeenCalled();
+      expect(component.isLoading).toBeFalse();
+      expect(notificationService.show).toHaveBeenCalledWith(
+        'Generation failed',
+        'error',
+        'cross-in-circle-white',
+        undefined,
+        20000,
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts:339]: fixed

File: frontend/src/app/audio/audio.component.ts [hunk 23]
```diff
@@ -258,8 +261,19 @@ export class AudioComponent {
...
+            error?.error?.detail?.[0]?.msg ||
+            error?.error?.detail ||
+            error?.message ||
+            'Something went wrong';
+          this.notificationService.show(
+            errorMessage,
+            'error',
+            'cross-in-circle-white',
+            undefined,
+            20000,
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.ts]: fixed

File: frontend/src/app/audio/audio.component.ts [hunk 24]
```diff
@@ -199,18 +197,26 @@ export class AudioComponent {
...
 
     const activeWorkspaceId = this.workspaceStateService.getActiveWorkspaceId();
     if (!activeWorkspaceId) {
-      handleErrorSnackbar(this.snackBar, { message: 'Please select a workspace first.' }, 'Workspace');
+      this.notificationService.show(
+        'Please select a workspace first.',
+        'error',
+        'cross-in-circle-white',
+        undefined,
+        20000,
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.ts]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 25]
```diff
@@ -0,0 +1,439 @@
...
+
+    it('should show error notification if no workspace is selected', () => {
+      workspaceStateService.getActiveWorkspaceId.and.returnValue(null);
+      component.generate();
+      expect(notificationService.show).toHaveBeenCalledWith(
+        'Please select a workspace first.',
+        'error',
+        'cross-in-circle-white',
+        undefined,
+        20000,
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 26]
```diff
@@ -0,0 +1,439 @@
...
+      imports: [
+        HttpClientTestingModule,
+        MatIconModule,
+        MockMediaLightboxComponent,
+      ],
+      providers: [
+        { provide: AudioService, useValue: audioServiceSpy },
+        { provide: WorkspaceStateService, useValue: workspaceStateServiceSpy },
+        {
+          provide: ActivatedRoute,
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 27]
```diff
@@ -0,0 +1,439 @@
...
+  GenerationModelEnum,
+} from '../services/audio/audio.service';
+import { of, throwError, Subject } from 'rxjs';
+import { JobStatus, MediaItem } from '../common/models/media-item.model';
+import { WorkspaceStateService } from '../services/workspace/workspace-state.service';
+import { MatSelectChange } from '@angular/material/select';
+import { MatIconModule } from '@angular/material/icon';
+import { LanguageEnum, VoiceEnum } from './audio.constants';
+import { By } from '@angular/platform-browser';
+import { ActivatedRoute } from '@angular/router';
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 28]
```diff
@@ -0,0 +1,465 @@
...
+    });
+
+    it('onAudioEnded should reset player state', () => {
+      component.isPlaying = true;
+      component.progressValue = 50;
+      component.currentTime = '1:00';
+      component.onAudioEnded();
+      expect(component.isPlaying).toBeFalse();
+      expect(component.progressValue).toBe(0);
+      expect(component.currentTime).toBe('0:00');
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts:428]: fixed added test case

File: frontend/src/app/audio/audio.component.spec.ts [hunk 29]
```diff
@@ -0,0 +1,457 @@
...
+      expect(audioService.generateAudio).toHaveBeenCalled();
+      expect(component.isLoading).toBeFalse();
+      expect(notificationService.show).toHaveBeenCalledWith(
+        'Backend validation error',
+        'error',
+        'cross-in-circle-white',
+        undefined,
+        20000,
+      );
+      tick(20000);
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: fixed

File: frontend/src/app/audio/audio.component.spec.ts (hunk 28 above)
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts:428]: case added

File: frontend/src/app/audio/audio.component.spec.ts [hunk 30]
```diff
@@ -0,0 +1,457 @@
...
+      expect(notificationService.show).toHaveBeenCalledWith(
+        'Backend validation error',
+        'error',
+        'cross-in-circle-white',
+        undefined,
+        5000,
+      );
+      tick(5000);
+    }));
+  });// Note: The above tests for the audio player are basic and can be expanded to cover more edge cases and interactions.
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: added new line

File: frontend/src/app/audio/audio.component.spec.ts [hunk 31]
```diff
@@ -0,0 +1,458 @@
...
+      expect(audioService.generateAudio).toHaveBeenCalled();
+      expect(component.isLoading).toBeFalse();
+      expect(notificationService.show).toHaveBeenCalledWith(
+        'Backend validation error',
+        'error',
+        'cross-in-circle-white',
+        undefined,
+        5000,
+      );
+      tick(5000);
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts]: removed

File: frontend/src/app/audio/audio.component.ts [hunk 32]
```diff
@@ -199,18 +197,26 @@ export class AudioComponent {
         };
         this.voices = [newVoice, ...this.voices];
         this.selectedVoice = newVoice.id;
-        handleSuccessSnackbar(this.snackBar, 'Voice cloned successfully!');
+        this.notificationService.show('Voice cloned successfully!', 'success', undefined, 'check_small');
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.ts]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 33]
```diff
@@ -0,0 +1,457 @@
...
+
+      component.openAddVoiceDialog();
+
+      expect(component.voices.length).toBe(initialVoiceCount + 1);
+      expect(notificationService.show).toHaveBeenCalledWith(
+        'Voice cloned successfully!',
+        'success',
+        undefined,
+        'check_small'
+      );
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts:465]: fixed

[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts:465]: fixed

File: frontend/src/app/audio/audio.component.ts [hunk 34]
```diff
@@ -199,18 +197,26 @@ export class AudioComponent {
...
-      handleErrorSnackbar(this.snackBar, { message: 'Please select a workspace first.' }, 'Workspace');
+      this.notificationService.show(
+        'Please select a workspace first.',
+        'error',
+        'cross-in-circle-white',
+        undefined,
+        20000
+      );
+      this.isLoading = false;
       return;
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.ts]: fixed

File: frontend/src/app/audio/audio.component.spec.ts [hunk 35]
```diff
@@ -0,0 +1,439 @@
...
+      expect(audioService.generateAudio).toHaveBeenCalled();
+      expect(component.isLoading).toBeFalse();
+      expect(notificationService.show).toHaveBeenCalledWith(
+        'Generation failed',
+        'error',
+        'cross-in-circle-white',
+        undefined,
+        20000,
+      );
+    }));
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts:364]: test case added

File: frontend/src/app/audio/audio.component.spec.ts [hunk 36]
```diff
@@ -0,0 +1,450 @@
...
+      expect(component.isPlaying).toBeTrue();
+    });
+
+    it('togglePlay should call pause() when playing', () => {
+      Object.defineProperty(audioEl, 'paused', { value: false });
+      component.togglePlay();
+      expect(audioEl.pause).toHaveBeenCalled();
+      expect(component.isPlaying).toBeFalse();
+    });
+
```
[the remark this answers is not in the document]
Comment (pradhyumnap-hub) [reply frontend/src/app/audio/audio.component.spec.ts:400]: after added this fix it will create many errors so unable to do
